### DISCUSSION TITLE
Solve optimistic concurrency issue without adding a `version` field

### DISCUSSION THREAD

**Discussion by Tsimopak** [prisma/prisma#24874] [Q&A]
### Question

In the following page there is a suggestion how to solve concurrency issue by adding a `version` field and ask about it in the `where` clause when updating a model: 

```
const userEmail = 'alice@prisma.io'
const movieName = 'Hidden Figures'

// Find the first available seat
// availableSeat.version might be 0
const availableSeat = await client.seat.findFirst({
  where: {
    Movie: {
      name: movieName,
    },
    claimedBy: null,
  },
})

if (!availableSeat) {
  throw new Error(`Oh no! ${movieName} is all booked.`)
}

// Only mark the seat as claimed if the availableSeat.version
// matches the version we're updating. Additionally, increment the
// version when we perform this update so all other clients trying
// to book this same seat will have an outdated version.
const seats = await client.seat.updateMany({
  data: {
    claimedBy: userEmail,
    version: {
      increment: 1,
    },
  },
  where: {
    id: availableSeat.id,
    version: availableSeat.version, // This version field is the key; only claim seat if in-memory version matches database version, indicating that the field has not been updated
  },
})

if (seats.count === 0) {
  throw new Error(`That seat is already booked! Please try again.`)
}
```
Reference: https://www.prisma.io/docs/orm/prisma-client/queries/transactions#optimistic-concurrency-control

what I want to ask, is if instead of doing:
```
  where: {
    id: availableSeat.id,
    version: availableSeat.version, // This version field is the key; only claim seat if in-memory version matches database version, indicating that the field has not been updated
  },
```

I could have asked if claimedBy is still null, will it do the same trick?
so the solution would be like so:
```
  where: {
    id: availableSeat.id,
    claimedBy: null
  },
```

The only disadvantage compared to version is that if there were many swaps between claimedby and null, i would have not known about it, but that's fine by me I guess?

**Comment by ludralph** [comment]
Hi @Tsimopak 👋 

Thank you for raising this question.

The `version` field is specifically designed for optimistic concurrency control. It ensures that the record has not been modified between the read and write operations by incrementing the version number with each update. The `version` field ensures that only one update can succeed by checking and incrementing the version atomically. It ensures that you are aware of any changes to the record between the time you read it and the time you attempt to update it. This method is robust because it detects any modification to the record, not just changes to the `claimedBy` field.

Using `claimedBy: null` does not provide the same level of concurrency control. It only checks if the seat is still unclaimed, but it does not account for any intermediate changes that might have occurred. If you use `claimedBy: null`, you might miss concurrent updates. For example, if two users try to claim the same seat simultaneously, both might pass the `claimedBy: null` check, leading to a race condition where the last update wins, potentially causing data inconsistency. The main disadvantage, as you mentioned, is that it won't detect if the seat was claimed and then released multiple times. This could lead to potential issues in scenarios where the seat's state changes frequently.

Here's the code using `claimedBy`:
```js
const userEmail = 'alice@prisma.io'
const movieName = 'Hidden Figures'

// Find the first available seat
const availableSeat = await client.seat.findFirst({
  where: {
    Movie: {
      name: movieName,
    },
    claimedBy: null,
  },
})

if (!availableSeat) {
  throw new Error(`Oh no! ${movieName} is all booked.`)
}

// Only mark the seat as claimed if it is still unclaimed
const seats = await client.seat.updateMany({
  data: {
    claimedBy: userEmail,
  },
  where: {
    id: availableSeat.id,
    claimedBy: null, // Check if the seat is still unclaimed
  },
})

if (seats.count === 0) {
  throw new Error(`That seat is already booked! Please try again.`)
}
```
The `findFirst` query looks for the first seat that matches the movie name and is unclaimed (`claimedBy: null`). If no available seat is found, it throws an error indicating that the movie is fully booked. The `updateMany` query attempts to update the seat to mark it as claimed by the user (`claimedBy: userEmail`), but only if it is still unclaimed (claimedBy: null). If no seats were updated (meaning the seat was claimed by someone else in the meantime), it throws an error indicating that the seat is already booked. 

This approach is simpler as it only checks if the seat is still unclaimed. However, it is slightly less robust compared to using a version field for concurrency control. It should work well for cases where the primary concern is to ensure the seat is unclaimed at the time of update. If there are frequent concurrent updates, you might still occasionally encounter race conditions. You can try wrapping the read and write operations in a database transaction to ensure atomicity,

**Reply by Tsimopak** [reply]
Is that a ChatGPT answer lol? I got the exact same answer there, you also repeated my code with no reason as ChatGPT does.
I'm looking for an human answer, thanks...
